File: backend/src/memory/Mem0Service.ts
/**
 * Mem0Service - Memory Management using Python service
 * 
 * This service communicates with a Python FastAPI service that handles
 * Mem0 operations, providing proper Azure OpenAI support.
 */

import axios, { AxiosInstance } from 'axios';
import { Message } from '../storage/PostgresSessionStorage';
import { STORAGE_CONFIG } from '../config/storage.config';

export interface Mem0Config {
    apiKey?: string;
    embeddingModel?: string;
    llmModel?: string;
    historyDbPath?: string;
    collectionName?: string;
    dimension?: number;
    provider?: string;
}

export interface MemorySearchOptions {
    limit?: number;
    userId?: number;
    sessionId?: string;
}

export interface MemoryAddResult {
    results: Array<{
        id: string;
        memory: string;
        metadata?: Record<string, any>;
    }>;
}

export interface MemorySearchResult {
    results: Array<{
        id: string;
        memory: string;
        score?: number;
        metadata?: Record<string, any>;
    }>;
}

export class Mem0Service {
    private apiClient: AxiosInstance;
    private initialized = false;
    private config: Mem0Config;
    private pythonServiceUrl: string;
    
    constructor(config?: Mem0Config) {
        this.pythonServiceUrl = process.env.MEM0_PYTHON_SERVICE_URL || 'http://localhost:8001';
        this.config = {
            apiKey: this.detectApiKey(),
            embeddingModel: this.detectEmbeddingModel(),
            llmModel: this.detectLLMModel(),
            historyDbPath: STORAGE_CONFIG.mem0.historyDbPath,
            collectionName: STORAGE_CONFIG.mem0.collectionName,
            dimension: this.detectDimension(),
            provider: STORAGE_CONFIG.mem0.provider,
            ...config
        };
        
        // Initialize axios client
        this.apiClient = axios.create({
            baseURL: this.pythonServiceUrl,
            timeout: 30000,
            headers: {
                'Content-Type': 'application/json'
            }
        });
    }

    private detectProvider(): 'azure' | 'openai' {
        const provider = (process.env.MEM0_MODELS || '').toLowerCase();
        console.log(`[Mem0Service] Detecting provider from MEM0_MODELS=${process.env.MEM0_MODELS}`);
        
        if (provider === 'azure') {
            console.log('[Mem0Service] Provider detected: azure');
            if (!process.env.AZURE_OPENAI_API_KEY || !process.env.AZURE_OPENAI_ENDPOINT) {
                console.error('[Mem0Service] Azure provider selected but missing required env vars');
                throw new Error('[Mem0Service] MEM0_MODELS=azure but AZURE_OPENAI_API_KEY or AZURE_OPENAI_ENDPOINT is missing.');
            }
            if (!process.env.AZURE_OPENAI_DEPLOYMENT || !process.env.AZURE_OPENAI_EMBEDDING_DEPLOYMENT) {
                console.error('[Mem0Service] Azure provider selected but missing deployment env vars');
                throw new Error('[Mem0Service] MEM0_MODELS=azure but AZURE_OPENAI_DEPLOYMENT or AZURE_OPENAI_EMBEDDING_DEPLOYMENT is missing.');
            }
            return 'azure';
        }
        if (provider === 'openai') {
            console.log('[Mem0Service] Provider detected: openai');
            if (!process.env.OPENAI_API_KEY) {
                console.error('[Mem0Service] OpenAI provider selected but OPENAI_API_KEY is missing');
                throw new Error('[Mem0Service] MEM0_MODELS=openai but OPENAI_API_KEY is missing.');
            }
            if (!process.env.MEM0_LLM_MODEL || !process.env.MEM0_EMBEDDING_MODEL) {
                console.error('[Mem0Service] OpenAI provider selected but missing model env vars');
                throw new Error('[Mem0Service] MEM0_MODELS=openai but MEM0_LLM_MODEL or MEM0_EMBEDDING_MODEL is missing.');
            }
            return 'openai';
        }
        console.error('[Mem0Service] Invalid MEM0_MODELS value:', process.env.MEM0_MODELS);
        throw new Error('[Mem0Service] MEM0_MODELS must be set to "openai" or "azure".');
    }

    private detectApiKey(): string {
        const provider = this.detectProvider();
        const apiKey = provider === 'azure' ? process.env.AZURE_OPENAI_API_KEY! : process.env.OPENAI_API_KEY!;
        console.log(`[Mem0Service] Using ${provider} API key (length: ${apiKey?.length || 0})`);
        return apiKey;
    }

    private detectEmbeddingModel(): string {
        const provider = this.detectProvider();
        const model = provider === 'azure' ? process.env.AZURE_OPENAI_EMBEDDING_DEPLOYMENT! : process.env.MEM0_EMBEDDING_MODEL!;
        console.log(`[Mem0Service] Using ${provider} embedding model: ${model}`);
        return model;
    }

    private detectLLMModel(): string {
        const provider = this.detectProvider();
        const model = provider === 'azure' ? process.env.AZURE_OPENAI_DEPLOYMENT! : process.env.MEM0_LLM_MODEL!;
        console.log(`[Mem0Service] Using ${provider} LLM model: ${model}`);
        return model;
    }

    private detectDimension(): number {
        // Both text-embedding-ada-002 and text-embedding-3-small use 1536 dims
        return 1536;
    }

    /**
     * Initialize the memory service
     */
    async initialize(): Promise<void> {
        if (this.initialized) {
            return;
        }
        
        try {
            // Check if Python service is healthy
            const health = await this.apiClient.get('/health');
            console.log('[Mem0Service] Python service health:', health.data);
            
            if (health.data.status === 'healthy' || health.data.status === 'initializing') {
                this.initialized = true;
                console.log('[Mem0Service] Connected to Python service successfully');
                
                // Get configuration info
                const config = await this.apiClient.get('/config');
                console.log('[Mem0Service] Python service config:', config.data);
            } else {
                throw new Error('Python service is not healthy');
            }
        } catch (error: any) {
            console.error('[Mem0Service] Failed to connect to Python service:', error.message);
            if (error.code === 'ECONNREFUSED') {
                console.error('[Mem0Service] Python service is not running. Start it with: cd python-mem0 && ./scripts/start.sh');
            }
            throw error;
        }
    }
    
    /**
     * Add messages to memory with user context
     */
    async addMessages(
        messages: Message[], 
        sessionId: string,
        userId?: number
    ): Promise<MemoryAddResult> {
        if (!this.initialized) {
            await this.initialize();
        }
        
        try {
            const response = await this.apiClient.post('/add', {
                messages: messages.map(msg => ({
                    id: msg.id,
                    type: msg.type,
                    content: msg.content,
                    timestamp: msg.timestamp
                })),
                session_id: sessionId,
                user_id: userId
            });
            
            console.log(`[Mem0Service] Added ${messages.length} messages to memory for session ${sessionId}, user ${userId}`);
            return response.data.results || { results: [] };
        } catch (error: any) {
            console.error('[Mem0Service] Failed to add messages:', error.message);
            throw error;
        }
    }
    
    /**
     * Add a single message to memory
     */
    async addMessage(
        message: Message,
        sessionId: string,
        userId?: number
    ): Promise<MemoryAddResult> {
        return this.addMessages([message], sessionId, userId);
    }
    
    /**
     * Search for relevant memories with user context
     */
    async search(
        query: string,
        options: MemorySearchOptions = {}
    ): Promise<MemorySearchResult> {
        if (!this.initialized) {
            await this.initialize();
        }
        
        try {
            const response = await this.apiClient.post('/search', {
                query: query,
                session_id: options.sessionId,
                user_id: options.userId,
                limit: options.limit || 10
            });
            
            console.log(`[Mem0Service] Search for "${query}" returned ${response.data.results?.length || 0} results`);
            return response.data;
        } catch (error: any) {
            console.error('[Mem0Service] Search failed:', error.message);
            throw error;
        }
    }
    
    /**
     * Get all memories for a session with user context
     */
    async getSessionMemories(
        sessionId: string,
        userId?: number,
        limit?: number
    ): Promise<any[]> {
        if (!this.initialized) {
            await this.initialize();
        }
        
        try {
            const response = await this.apiClient.post('/get-session-memories', {
                session_id: sessionId,
                user_id: userId,
                limit: limit || 100
            });
            
            const memories = response.data.memories || [];
            console.log(`[Mem0Service] Retrieved ${memories.length} memories for session ${sessionId}`);
            return memories;
        } catch (error: any) {
            console.error('[Mem0Service] Failed to get session memories:', error.message);
            throw error;
        }
    }
    
    /**
     * Get context messages for LLM with user awareness
     */
    async getContextForQuery(
        query: string,
        sessionId: string,
        userId?: number,
        maxMessages: number = 50
    ): Promise<Array<{role: string, content: string}>> {
        if (!this.initialized) {
            await this.initialize();
        }
        
        try {
            const response = await this.apiClient.post('/get-context', {
                query: query,
                session_id: sessionId,
                user_id: userId,
                max_messages: maxMessages
            });
            
            const contextMessages = response.data.context_messages || [];
            console.log(`[Mem0Service] Built context with ${contextMessages.length} relevant memories`);
            return contextMessages;
        } catch (error: any) {
            console.error('[Mem0Service] Failed to get context:', error.message);
            throw error;
        }
    }
    
    /**
     * Delete memories for a session with user context
     */
    async deleteSessionMemories(
        sessionId: string, 
        userId?: number
    ): Promise<void> {
        if (!this.initialized) {
            await this.initialize();
        }
        
        try {
            const response = await this.apiClient.post('/delete-session-memories', {
                session_id: sessionId,
                user_id: userId
            });
            
            console.log(`[Mem0Service] Deleted ${response.data.deleted_count || 0} memories for session ${sessionId}`);
        } catch (error: any) {
            console.error('[Mem0Service] Failed to delete session memories:', error.message);
            throw error;
        }
    }
    
    /**
     * Get memory history for a session with user context
     */
    async getMemoryHistory(
        sessionId: string,
        userId?: number
    ): Promise<any[]> {
        // This method is not implemented in the Python service yet
        console.warn('[Mem0Service] getMemoryHistory is not implemented in Python service');
        return [];
    }
    
    /**
     * Get all memories for a user across all sessions
     */
    async getAllUserMemories(userId: number, limit: number = 1000): Promise<any[]> {
        if (!this.initialized) {
            await this.initialize();
        }

        try {
            const response = await this.apiClient.post('/get-all-user-memories', {
                user_id: userId,
                limit: limit
            });

            const memories = response.data.memories || [];
            console.log(`[Mem0Service] Retrieved ${memories.length} total memories for user ${userId}`);
            return memories;

        } catch (error: any) {
            console.error('[Mem0Service] Failed to get all user memories:', error.message);
            return [];
        }
    }

    /**
     * Get user statistics (not implemented in Python service)
     */
    async getUserMemoryStats(userId: number): Promise<any> {
        // This would need to be implemented in the Python service
        return null;
    }
}

// Singleton instance
let mem0ServiceInstance: Mem0Service | null = null;

/**
 * Get or create the Mem0Service instance
 */
export function getMem0Service(config?: Mem0Config): Mem0Service {
    if (!mem0ServiceInstance) {
        mem0ServiceInstance = new Mem0Service(config);
    }
    return mem0ServiceInstance;
}

File: backend/python-mem0/src/mem0_service.py
"""Mem0 service wrapper for handling memory operations."""
import logging
from typing import List, Dict, Any, Optional
from mem0 import Memory
from .config import config

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class Mem0Service:
    """Wrapper for Mem0 memory operations."""
    
    def __init__(self):
        self.config = config
        self.memory: Optional[Memory] = None
        self.initialized = False
    
    def initialize(self) -> None:
        """Initialize the Mem0 memory instance."""
        if self.initialized:
            return
        
        try:
            logger.info(f"Initializing Mem0 with provider: {self.config.provider}")
            mem0_config = self.config.get_mem0_config()
            
            # Log configuration (without sensitive data)
            logger.info(f"Using embedding model: {self.config.embedding_model}")
            logger.info(f"Using LLM model: {self.config.llm_model}")
            if self.config.provider == "azure":
                logger.info(f"Azure endpoint: {self.config.azure_endpoint}")
                logger.info(f"Azure API version: {self.config.azure_api_version}")
            
            # Initialize Mem0
            self.memory = Memory.from_config(mem0_config)
            self.initialized = True
            logger.info("Mem0 initialized successfully")
            
        except Exception as e:
            logger.error(f"Failed to initialize Mem0: {str(e)}")
            raise
    
    async def add_messages(
        self, 
        messages: List[Dict[str, Any]], 
        session_id: str, 
        user_id: Optional[int] = None
    ) -> Dict[str, Any]:
        """Add messages to memory."""
        if not self.initialized:
            self.initialize()
        
        try:
            # Convert message format if needed
            mem0_messages = []
            for msg in messages:
                role = "user" if msg.get("type") == "user" else "assistant"
                mem0_messages.append({
                    "role": role,
                    "content": msg.get("content", "")
                })
            
            # Build metadata
            metadata = {
                "session_id": session_id,
                "user_id": str(user_id) if user_id else "default"
            }
            
            # Add to memory
            result = self.memory.add(
                mem0_messages,
                user_id=str(user_id) if user_id else "default",
                metadata=metadata
            )
            
            logger.info(f"Added {len(messages)} messages to memory for session {session_id}")
            return {"success": True, "results": result}
            
        except Exception as e:
            logger.error(f"Failed to add messages: {str(e)}")
            return {"success": False, "error": str(e), "results": []}
    
    async def search(
        self, 
        query: str, 
        session_id: Optional[str] = None,
        user_id: Optional[int] = None,
        limit: int = 10
    ) -> Dict[str, Any]:
        """Search for relevant memories."""
        if not self.initialized:
            self.initialize()
        
        try:
            search_params = {
                "query": query,
                "user_id": str(user_id) if user_id else "default",
                "limit": limit
            }
            
            # Add session filter if provided
            if session_id:
                search_params["filters"] = {"session_id": session_id}
            
            results = self.memory.search(**search_params)
            
            # Debug log
            logger.info(f"Raw search results type: {type(results)}")
            
            # Handle different response formats from mem0
            if isinstance(results, dict):
                # If it's a dict with 'results' key, extract the results
                actual_results = results.get('results', [])
                logger.info(f"Got dict with 'results' key, extracting {len(actual_results)} results")
            elif isinstance(results, list):
                actual_results = results
                logger.info(f"Got list directly with {len(actual_results)} results")
            else:
                logger.warning(f"Unexpected results format: {type(results)}")
                actual_results = []
            
            # Normalize results to always be dicts
            normalized_results = []
            for idx, result in enumerate(actual_results):
                if isinstance(result, dict):
                    normalized_results.append(result)
                elif isinstance(result, str):
                    normalized_results.append({
                        "id": str(idx),
                        "memory": result,
                        "score": 1.0
                    })
            
            logger.info(f"Search for '{query}' returned {len(normalized_results)} results")
            return {"success": True, "results": normalized_results}
            
        except Exception as e:
            logger.error(f"Search failed: {str(e)}")
            return {"success": False, "error": str(e), "results": []}
    
    async def get_all_user_memories(
        self,
        user_id: int,
        limit: int = 1000
    ) -> List[Dict[str, Any]]:
        """Get all memories for a user across all sessions."""
        if not self.initialized:
            self.initialize()
        
        try:
            all_memories = self.memory.get_all(
                user_id=str(user_id) if user_id else "default",
                limit=limit
            )
            
            # Debug log
            logger.info(f"Raw all_memories type: {type(all_memories)}")
            
            # Handle different response formats
            if isinstance(all_memories, dict):
                actual_memories = all_memories.get('results', all_memories.get('memories', []))
                logger.info(f"Got dict response, extracted {len(actual_memories)} memories")
            elif isinstance(all_memories, list):
                actual_memories = all_memories
                logger.info(f"Got list response with {len(actual_memories)} memories")
            else:
                logger.warning(f"Unexpected all_memories format: {type(all_memories)}")
                actual_memories = []
            
            # Return all memories without filtering by session
            formatted_memories = []
            for mem in actual_memories:
                if isinstance(mem, dict):
                    formatted_memories.append(mem)
                elif isinstance(mem, str):
                    formatted_memories.append({
                        "id": str(len(formatted_memories)),
                        "memory": mem,
                        "metadata": {}
                    })
            
            logger.info(f"Retrieved {len(formatted_memories)} total memories for user {user_id}")
            return formatted_memories
            
        except Exception as e:
            logger.error(f"Failed to get all user memories: {str(e)}")
            return []
    
    async def get_session_memories(
        self, 
        session_id: str, 
        user_id: Optional[int] = None,
        limit: int = 100
    ) -> List[Dict[str, Any]]:
        """Get all memories for a session."""
        if not self.initialized:
            self.initialize()
        
        try:
            all_memories = self.memory.get_all(
                user_id=str(user_id) if user_id else "default",
                limit=limit
            )
            
            # Debug log to see what we're getting
            logger.info(f"Raw all_memories type: {type(all_memories)}")
            
            # Handle different response formats
            if isinstance(all_memories, dict):
                # If it's a dict with 'results' or 'memories' key, extract them
                actual_memories = all_memories.get('results', all_memories.get('memories', []))
                logger.info(f"Got dict response, extracted {len(actual_memories)} memories")
            elif isinstance(all_memories, list):
                actual_memories = all_memories
                logger.info(f"Got list response with {len(actual_memories)} memories")
            else:
                logger.warning(f"Unexpected all_memories format: {type(all_memories)}")
                actual_memories = []
            
            # Filter by session_id - handle both dict and string responses
            session_memories = []
            for mem in actual_memories:
                if isinstance(mem, dict):
                    # Check if this memory belongs to the session
                    mem_session_id = mem.get("metadata", {}).get("session_id")
                    if mem_session_id == session_id:
                        session_memories.append(mem)
                elif isinstance(mem, str):
                    # If it's a string, wrap it in a dict format
                    session_memories.append({
                        "id": str(len(session_memories)),
                        "memory": mem,
                        "metadata": {"session_id": session_id}
                    })
            
            logger.info(f"Retrieved {len(session_memories)} memories for session {session_id}")
            return session_memories
            
        except Exception as e:
            logger.error(f"Failed to get session memories: {str(e)}")
            return []
    
    async def delete_session_memories(
        self, 
        session_id: str, 
        user_id: Optional[int] = None
    ) -> Dict[str, Any]:
        """Delete all memories for a session."""
        if not self.initialized:
            self.initialize()
        
        try:
            # Get all memories for the session
            memories = await self.get_session_memories(session_id, user_id)
            
            # Delete each memory
            deleted_count = 0
            for memory in memories:
                if memory.get("id"):
                    self.memory.delete(memory["id"])
                    deleted_count += 1
            
            logger.info(f"Deleted {deleted_count} memories for session {session_id}")
            return {"success": True, "deleted_count": deleted_count}
            
        except Exception as e:
            logger.error(f"Failed to delete session memories: {str(e)}")
            return {"success": False, "error": str(e), "deleted_count": 0}
    
    async def get_context_for_query(
        self,
        query: str,
        session_id: str,
        user_id: Optional[int] = None,
        max_messages: int = 50
    ) -> List[Dict[str, str]]:
        """Get context messages for LLM based on query."""
        if not self.initialized:
            self.initialize()
        
        try:
            # Search for relevant memories (cross-session for user context)
            search_results = await self.search(query, user_id=user_id, limit=max_messages)
            
            # Convert to context messages
            context_messages = []
            for result in search_results.get("results", []):
                memory_content = result.get("memory", "")
                if memory_content:
                    context_messages.append({
                        "role": "system",
                        "content": f"[Relevant Context: {memory_content}]"
                    })
            
            logger.info(f"Built context with {len(context_messages)} relevant memories")
            return context_messages
            
        except Exception as e:
            logger.error(f"Failed to get context: {str(e)}")
            return []


# Global service instance
mem0_service = Mem0Service()